Detections/SigninLogs/AnomalousIPUsageFollowedByTeamsAction.kql

```diff
@@ -0,0 +1,104 @@
+  | join (
+      //Get all successful login events to teams from each country
+      SigninLogs
+      | where TimeGenerated > ago(timeRange)
+      | where AppDisplayName contains "Teams"
+      | where ConditionalAccessStatus =~ "success"
+      | extend country = LocationDetails['countryOrRegion']
+      | where country != ""
+      | summarize count() by tostring(country)
+      | join (
+          //Now get the total number of logins from any country and join it to the previous count in a single table
+          SigninLogs
+          | where TimeGenerated > ago(timeRange)
+          | where AppDisplayName contains "Teams"
+          | where ConditionalAccessStatus =~ "success"
+          | extend country = LocationDetails['countryOrRegion']
+          | where country != ""
+          | summarize count(), make_list(tostring(country))
+          | mv-expand list_country
+          | extend country = tostring(list_country)
+      ) on country
+      | summarize by country, count_, count_1
+      //Now calculate each countries prevalence within login events
+      | extend prevalence = toreal(count_) / toreal(count_1) * 100
+      | project-away count_1
+  ) on country
+  //Login start and end times from the JSON object, this is the activity window the suspicious IP was active within
+  | extend LoginStartTime = make_datetime(tostring(bag_p.[tostring(min_count_)].["Start"]))
+  | extend LoginEndTime = make_datetime(tostring(bag_p.[tostring(min_count_)].["End"]))
+  | extend SuspiciousIP = tostring(bag_p.[tostring(min_count_)].["IP"])
+  | project UserPrincipalName, SuspiciousIP, UserIPs = bag_p, IPCountryMapping = set_any_country, UserIPDelta = delta, SuspiciousLoginCountry = country, SuspiciousCountryPrevalence = prevalence, LoginStartTime, LoginEndTime 
+  //The % that suspicious country is prevalent in data, this can be configured, less than 10% is uncommon
+  | where SuspiciousCountryPrevalence < countryPrevalenceThreshold
+  //Teams join to collect operations the user account has performed within the given time range
+  | join kind=inner( 
+  TeamsData 
+  | where TimeGenerated >= ago(timeRange) 
+  | project Operation, UserId=tolower(UserId), OperationTime=TimeGenerated
+  ) on $left.UserPrincipalName == $right.UserId
+  //The end time is projected 60 minutes forward, in case actions took place within the last hour of the final login for the suspicious IP
+  | extend ProjectedEndTime = LoginEndTime + projectedEndTime
+  //Limit to operations carried out by the user account in the timeframe the IP was active
+  | where OperationTime between (LoginStartTime .. ProjectedEndTime)
+  //Filter on suspicious actions
+  | project UserPrincipalName, SuspiciousIP, IPCountryMapping, LoginStartTime, LoginEndTime, Operation, IPCustomEntity=SuspiciousIP, AccountCustomEntity=UserPrincipalName
+  | where Operation in~ ("TeamsAdminAction", "MemberAdded", "MemberRoleChanged", "AppInstalled", "BotAddedToTeam")
```

Comment:
move this filter into the join of TeamsData?  Otherwise you are processing all the TeamsData for no real use.

Reply:
This has been moved now